src/Nethermind/Nethermind.Core/Crypto/KeccakCache.cs

```diff
@@ -0,0 +1,217 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Threading;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Core.Crypto;
+
+/// <summary>
+/// This is a minimalistic one-way set associative cache for Keccak values.
+///
+/// It allocates only 12MB of memory to store 128k of entries.
+/// No misaligned reads. Everything is aligned to both cache lines as well as to boundaries so no torn reads.
+/// Requires a single CAS to lock and <see cref="Volatile.Write(ref int,int)"/> to unlock.
+/// On a lock failure, it just moves on with execution.
+/// When a potential hit happens, the value of the result and the value of the key are copied on the stack to release the lock ASAP.
+/// </summary>
+public static unsafe class KeccakCache
+{
+    /// <summary>
+    /// Count is defined as a +1 over bucket mask. In the future, just change the mask as the main parameter.
+    /// </summary>
+    public const int Count = BucketMask + 1;
+
+    private const int BucketMask = 0x0001_FFFF;
+    private const uint HashMask = unchecked((uint)~BucketMask);
+    private const uint LockMarker = 0x0000_8000;
+
+    private const int InputLengthOfKeccak = ValueHash256.MemorySize;
+    private const int InputLengthOfAddress = Address.Size;
+
+    private static readonly Entry* Memory;
+
+    static KeccakCache()
+    {
+        const UIntPtr size = Count * Entry.Size;
+
+        // Aligned, so that no torn reads if fields of Entry are properly aligned.
+        Memory = (Entry*)NativeMemory.AlignedAlloc(size, BitOperations.RoundUpToPowerOf2(Entry.Size));
+        NativeMemory.Clear(Memory, size);
+        GC.AddMemoryPressure((long)size);
+    }
+
+    [SkipLocalsInit]
+    public static ValueHash256 Compute(ReadOnlySpan<byte> input)
+    {
+        ComputeTo(input, out ValueHash256 keccak256);
+        return keccak256;
+    }
+
+    [SkipLocalsInit]
+    public static void ComputeTo(ReadOnlySpan<byte> input, out ValueHash256 keccak256)
+    {
+        // Special cases jump forward as unpredicted
+        if (input.Length == 0 || input.Length > Entry.MaxPayloadLength)
+        {
+            goto Uncommon;
+        }
+
+        int hashCode = input.FastHash();
+        uint index = (uint)hashCode & BucketMask;
```

Comment:
So it will hold only one hash with given prefix?

Reply:
It will hold one hash per bucket, where the bucket is determined by a `FastHash` over the `input` data. On collision over `FastHash` it tries to lock and exchange (using `Interlocked`) but if it fails it just moves on. This is to make it as cheap as possible for both contended and not contended buckets.